### PR TITLE
Improve SSBO support for multiple & array structs

### PR DESCRIPTION
**Pull request by ricardomatias**
You can declare f.ex:
```kotlin
val block1 = shaderStorageBuffer(shaderStorageFormat {
    member("trap", BufferMemberType.VECTOR2_FLOAT, 8)
    struct("Particle", "particles", 4) {
        member("position", BufferMemberType.VECTOR3_INT)
        member("velocity", BufferMemberType.VECTOR3_FLOAT)
        member("age", BufferMemberType.FLOAT)
        member("isActive", BufferMemberType.BOOLEAN)
    }
    struct("Agent", "agents", 2) {
        member("position", BufferMemberType.VECTOR3_FLOAT)
    }
})
```

which would be used in GLSL f.ex:
```kotlin
struct Particle {
    ivec3 position;
    vec3 velocity;
    float age;
    bool isActive;
};

struct Agent {
    vec3 position;
};

layout (std430, binding = 0) buffer Block1 {
    vec2 trap[8];
    Particle particles[4];
    Agent agents[2];
} block1;
```

### CODE REVIEW

File: openrndr-gl3/src/main/kotlin/org/openrndr/internal/gl3/BufferWriterStd430GL3.kt [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+                if (pointer >= (index + element.arraySize)) {
+                    pointer = index + 1
+                    element = elements[pointer]
+                }
+            }
+            is ShaderStorageStruct -> {
+                val struct = element as ShaderStorageStruct
+                val index = elements.indexOf(element)
+
+                // pointer = 2
```
Comment (ricardomatias) [inline openrndr-gl3/src/main/kotlin/org/openrndr/internal/gl3/BufferWriterStd430GL3.kt:41]: to remove

File: openrndr-gl3/src/main/kotlin/org/openrndr/internal/gl3/ComputeShaderGL43.kt [hunk 2]
```diff
@@ -25,16 +25,16 @@ class ComputeShaderGL43(val programObject: Int, val name: String = "compute_shad
...
     private val uniforms: MutableMap<String, Int> = hashMapOf()
     private fun uniformIndex(uniform: String, query: Boolean = false): Int =
-            uniforms.getOrPut(uniform) {
-                val location = glGetUniformLocation(programObject, uniform)
-                debugGLErrors()
-                if (location == -1 && !query) {
-                    logger.warn {
-                        "Shader '${name}' does not have a uniform called '$uniform'"
-                    }
+        uniforms.getOrPut(uniform) {
```
Comment (ricardomatias) [inline openrndr-gl3/src/main/kotlin/org/openrndr/internal/gl3/ComputeShaderGL43.kt:28]: format related change

File: openrndr-gl3/src/main/kotlin/org/openrndr/internal/gl3/ComputeShaderGL43.kt [hunk 3]
```diff
@@ -64,7 +64,15 @@ class ComputeShaderGL43(val programObject: Int, val name: String = "compute_shad
                 require(colorBuffer.format.componentCount != 3) {
                     "color buffer has unsupported format (${imageBinding.colorBuffer.format}), only formats with 1, 2 or 4 components are supported"
                 }
-                GL43C.glBindImageTexture(image, colorBuffer.texture, 0, false, 0, imageBinding.access.gl(), colorBuffer.glFormat())
+                GL43C.glBindImageTexture(
```
Comment (ricardomatias) [inline openrndr-gl3/src/main/kotlin/org/openrndr/internal/gl3/ComputeShaderGL43.kt:67]: format related change

File: openrndr-gl3/src/main/kotlin/org/openrndr/internal/gl3/ComputeShaderGL43.kt [hunk 4]
```diff
@@ -94,6 +101,26 @@ class ComputeShaderGL43(val programObject: Int, val name: String = "compute_shad
         }
 
 
+    }
+
+    override fun buffer(name: String, shaderStorageBuffer: ShaderStorageBuffer) {
+        val index = storageIndex.getOrPut(name) { storageIndex.size + 8 }
+
+        println("shaderStorageBuffer index: $index")
```
Comment (ricardomatias) [inline openrndr-gl3/src/main/kotlin/org/openrndr/internal/gl3/ComputeShaderGL43.kt:109]: to remove

**Comment by shi-weili** [comment]
This will be so much nicer than using VBO as SSBO! 🙌

**Pull request merged**